File: src/main/java/com/github/simiacryptus/aicoder/util/psi/PsiVisitorBase.java
package com.github.simiacryptus.aicoder.util.psi;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.atomic.AtomicReference;

public abstract class PsiVisitorBase {

    public PsiVisitorBase() {
    }

    public void build(@NotNull PsiFile psiFile) {
        @NotNull AtomicReference<PsiElementVisitor> visitor = new AtomicReference<>();
        visitor.set(new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                visit(element, visitor.get());
                super.visitElement(element);
            }
        });
        psiFile.accept(visitor.get());
    }

    protected abstract void visit(@NotNull PsiElement element, PsiElementVisitor self);
}

File: src/main/kotlin/com/simiacryptus/aicoder/util/psi/PsiVisitorBase.kt
package com.simiacryptus.aicoder.util.psi

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.PsiFile
import java.util.concurrent.atomic.AtomicReference

abstract class PsiVisitorBase {
  fun build(psiFile: PsiFile) {
    val visitor = AtomicReference<PsiElementVisitor>()
    visitor.set(object : PsiElementVisitor() {
      override fun visitElement(element: PsiElement) {
        visit(element, visitor.get())
        super.visitElement(element)
      }
    })
    psiFile.accept(visitor.get())
  }

  protected abstract fun visit(element: PsiElement, self: PsiElementVisitor)
}
